src/lib/Wst/Offchain/Query.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-| Look up outputs at script addresses
+-}
+module Wst.Offchain.Query(
+  -- * Queries
+  blacklistNodes,
+  registryNodes,
+  globalParamsNode,
+  programmableLogicOutputs,
+  userProgrammableOutputs,
+
+  -- * UTxO with datum
+  UTxODat(..),
+  fromOutput
+) where
+
+import Cardano.Api qualified as C
+import Control.Lens qualified as L
+import Control.Monad ((>=>))
+import Control.Monad.Except (MonadError, throwError)
+import Control.Monad.Reader (MonadReader, asks)
+import Convex.CardanoApi.Lenses qualified as L
+import Convex.Class (MonadBlockchain (queryNetworkId), MonadUtxoQuery,
+                     utxosByPaymentCredential)
+import Convex.PlutusLedger.V1 (transCredential, unTransStakeCredential)
+import Convex.Scripts (fromHashableScriptData)
+import Convex.Utxos (UtxoSet, toApiUtxo)
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Map qualified as Map
+import Data.Maybe (listToMaybe, mapMaybe)
+import GHC.Generics (Generic)
+import PlutusTx qualified
+import SmartTokens.Types.ProtocolParams (ProgrammableLogicGlobalParams)
+import SmartTokens.Types.PTokenDirectory (BlacklistNode, DirectorySetNode (..))
+import Wst.AppError (AppError (GlobalParamsNodeNotFound))
+import Wst.Offchain.Env (DirectoryEnv (dsDirectorySpendingScript, dsProgrammableLogicBaseScript),
+                         HasDirectoryEnv (directoryEnv),
+                         HasTransferLogicEnv (transferLogicEnv),
+                         TransferLogicEnv (tleBlacklistSpendingScript))
+import Wst.Offchain.Scripts (protocolParamsSpendingScript)
+
+-- TODO: We should probably filter the UTxOs to check that they have the correct NFTs
+
+{-| Unspent transaction output with 'TxIn', 'TxOut' and an inline datum
+-}
+data UTxODat era a =
+  UTxODat
+    { uIn    :: C.TxIn
+    , uOut   :: C.TxOut C.CtxUTxO era
+    , uDatum :: a
+    }
+  deriving stock (Eq, Show, Generic)
+  deriving anyclass (ToJSON, FromJSON)
+
+{-| Find all UTxOs that make up the registry
+-}
+registryNodes :: forall era env m. (MonadReader env m, HasDirectoryEnv env, MonadUtxoQuery m, C.IsBabbageBasedEra era) => m [UTxODat era DirectorySetNode]
+registryNodes =
+  asks (C.PaymentCredentialByScript . C.hashScript . C.PlutusScript C.PlutusScriptV3 . dsDirectorySpendingScript . directoryEnv)
+    >>= fmap (extractUTxO @era) . utxosByPaymentCredential
```

Comment:
This is not a sufficient query, it should either query `utxosByPolicyId` or query `utxosByPaymentCredential` and then filter the result such that it returns only UTxOs that contain a token with the directoryMintingScript PolicyId